Add support for Japan to the tax details modal

haszari opened the pull request:
Fixes #8926

#### Changes proposed in this Pull Request
This PR changes the tax details modal so it works well for merchants in Japan, alongside existing VAT merchants in EU and UK.

Prior to this PR, the modal assumes `VAT` and various other details of the flow are tailored to the currently-supported tax regions:

- Use of "VAT" throughout
- Specific hints for formatting of the VAT number
- Business address details are powered by the validation API (at least for some regions, e.g. UK)

To make the modal work for Japan, I've generalised it a little, by parameterising all the data/copy/content that is region-specific:

- The name of the tax ID is factored out (e.g. `VAT number` vs `Corporate Number`; in future `GST number`).
- The name of the tax is factored out (e.g. `VAT` vs `Consumption Tax`; in future `GST`).
- The validation hint is customisable per region.

I've made some other changes so things are clear and accurate

- Generalised the title, and made it more friendly/clear - now `Set your tax details` (covers all tasks - id and address).
- Fixed mention of `receipts` to `tax documents`, more consistent with rest of the UI and documentation.

##### here's what's changed, in visual form
In the first screenshot, I've highlighted the non-essential stuff I've changed.

| Step | Old - UK | New - UK | New – Japan |
| --- | --- | --- | --- |
| Open modal | <img width="764" alt="Existing UK 1" src="https://github.com/user-attachments/assets/98df3bf4-0b39-411b-a487-6cb7b0e365d5"> | <img width="729" alt="New UK 1" src="https://github.com/user-attachments/assets/97e123db-3737-4577-873f-395e79421542"> | <img width="680" alt="JP 1" src="https://github.com/user-attachments/assets/96cebb66-138a-48c7-b5fc-e9a83a6ac95c"> |
| Bad tax ID † | <img width="738" alt="Existing UK 2 - invalid VAT ID" src="https://github.com/user-attachments/assets/4f0b1fab-772b-4fd8-a207-1850b8e11a94"> | <img width="728" alt="New UK 2 - invalid" src="https://github.com/user-attachments/assets/126feef2-08d6-4267-b2c4-ee69f2f9c846"> | <img width="676" alt="JP 2 - invalid ID" src="https://github.com/user-attachments/assets/29cf1db1-ffa7-4015-b729-932218a90c82"> |
| Confirm details | <img width="725" alt="Existing UK 3 - enter details, valid ID" src="https://github.com/user-attachments/assets/67cadeb8-6598-44f2-ab44-5137afb773b1"> | <img width="726" alt="New UK 3 - details" src="https://github.com/user-attachments/assets/775d1baa-bf91-4619-9a78-d0e13362fb86"> | <img width="677" alt="JP 3 - valid, enter details" src="https://github.com/user-attachments/assets/06bd3ec1-5d06-4bcf-95ba-a6223ff9c7ce"> |

@rogermattic keen for your feedback on this, if you see anything amiss please comment!

† Note: The error message for a bad corporate number / Japan tax ID is currently incorrect. I've made server changes to allow us to customise these error messages, I'm proposing doing that in a follow up PR to keep this PR moving. 

- https://github.com/Automattic/woocommerce-payments/issues/9144


##### this is arguably a temporary fix
IMO this dialog is quite convoluted and complex, since each region is subtly different.

Long term, I think we should implement separate modals/components for each region, with a similar interface, and have one outer conditional for region. I didn't undertake that on this PR because I wanted a relatively quick and low-risk fix. 

Why: If each region's UI is self-contained, it's easier to review and test, or customise the flow if needed. Also IMO it's not good practice to translate substrings such as `VAT number`, it's better to translate whole sentences (or ideally whole modal/UI).

Next time we add a new region, or do any work, let's consider that approach. _Reviewers:_ keen for feedback on this aspect specifically.

#### Testing instructions
Note you'll need to test this with a store with [account country that supports tax docs](https://github.com/Automattic/woocommerce-payments-server/blob/8f6bdc2195cd0a99abf888300b292795e7d294c7/server/wp-content/rest-api-plugins/endpoints/wcpay/utils/class-vat-utils.php#L19). Ideally, test with different accounts to see UI for different countries. Tip: if you set `wcpay_accounts[].deleted` to a date in the past, you can onboard a new account. Or switch between accounts by setting wcpay_accounts[].deleted=NULL` on the account you want to use.

1. Set up tax docs so you have a document to view at `Payments > Documents`. Fake one up with this CLI command:
   - `wp wcpay add_document_for_account [account_id] vat_invoice --mode=test --period_from="2024-04-01 00:00:00" --period_to="2024-04-30 23:59:59"`
   - Or can use `wp wcpay backfill_merchant_vat_invoices`
1. If you've added VAT id / tax information previously, clear it out so you can see the modal: 
   - Delete `company_vat_submitted` `company_vat_number` `company_name` `company_address` from  `wcpay_account_meta` table.
   - Side note:
       - https://github.com/Automattic/woocommerce-payments-server/issues/5915
1. Go to `Payments > Documents`. 
1. Click `Download` for a tax doc row.
1. See modal.
1. Test adding valid and invalid tax info.
1. Test downloading the generated tax doc – confirm the merchant details are correct.


-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _not requesting GlobalStep testing, will need Japan store onboarded to test with, better to just test ourselves_
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable. _not affecting a critical flow; tax docs is a secondary feature_
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Commit:
generalise title of VAT/GST modal

Commit:
WIP/experiment: refactor VAT modal with country specific hint & tax name

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8974` or branch name `fix/8926-localise-vat-details-modal` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/8926-localise-vat-details-modal) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: bfb3bd110a5783a991d8a4795ac39573d0766f34
- Build time: 2024-07-25 03:00:24 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Commit:
tweak title to align more closely with existing user docs

Commit:
factor out "you gotta register a VAT number" hint into a country-specific getter, since it might not work well for Japan

Commit:
comments / ideas for requirement hint

Commit:
parameterise most mentions of "VAT number", remove enforced prefix for Japan

Commit:
remove getVatTaxName, not used

Commit:
comment to explain that JP has no prefix

haszari committed:
Merge branch 'develop' into fix/8926-localise-vat-details-modal

haszari committed:
Merge branch 'develop' into fix/8926-localise-vat-details-modal

haszari commented on client/vat/form-modal/index.tsx:
```diff
@@ -24,7 +24,7 @@ const VatFormModal = ( {
 	return isModalOpen ? (
 		<Modal
-			title={ __( 'VAT details', 'woocommerce-payments' ) }
+			title={ __( 'Set your tax details', 'woocommerce-payments' ) }
```
Note – I've changed the dialog title for two reasons:

1. Remove `VAT`, not all regions call it `VAT`.
2. Generalise, higher level title. 

Re (2)…

The dialog's purpose is to set various tax details – the tax/VAT number, and address details. It's a multi-step process, so I think this title is a clearer summary of both steps.

Commit:
document VatNumberTask component, remove TODO comment

Commit:
add comment suggestions for alternative long-term approach: component per region

Commit:
changelog

haszari commented:
> Covered with tests

FYI this is ready for review, but I still need to fix up the e2e tests. Reviewers please let me know if you have suggestions for scenarios to e2e test!

shendy-a8c committed:
Merge branch 'develop' into fix/8926-localise-vat-details-modal

Commit:
add relevant japan tax keywords to changelog

haszari committed:
Merge branch 'develop' into fix/8926-localise-vat-details-modal

Commit:
update tax modal e2e tests:
- fix expected copy – old code was using unusual apostrophe char
- expect no prefix for JP corporate number

Commit:
update various tests for renamed "Tax details" form

Commit:
reduce specificity of test, `VAT number` is not hard-coded in UI anymore

haszari commented:
Note for reviewers: there are a bunch of tests that need updating/fixing, I'm in progress on this :) 

Feedback or suggestions for improving the tests is welcome!

shendy-a8c commented on client/vat/form/tasks/vat-number-task.tsx:
```diff
@@ -33,6 +39,54 @@ const getVatPrefix = () => {
 	}
 };
 
+const getVatTaxIDName = () => {
```
The return value of `getVatTaxIDName()` is the return value of `__()` but I see it will be fed into another `__()`. I was concerned that the string will be double-translated, but I think it will not because it's considered as a parameter for the 2nd `__()` call.

haszari replied:
This is part of the reason why I would prefer to have separate dialogs for each region (as needed). Then the translator can translate as whole, and [avoid concatenating translated strings (which is a localisation no-no)](https://sapegin.me/blog/internationalization).

brucealdridge replied:
I think the way it's done here is okay. We're not doing `{phrase1} {tax number} {phrase2}` we're instead doing `Set your %s` (where %s is `tax_number`) which is exactly how that link recommends doing it.
While whole strings would be better, this approach is fine.

haszari replied:
Yea it's shippable as is, but the amount of branching and substitutions is the problem, _() on _() is a symptom of that.

> While whole strings would be better, this approach is fine.

Exactly, whole strings (or UIs) would be better, my plan is next time we work on this we do that refactor.

shendy-a8c reviewed (approved):
Looks good and tested well.

Thank you for the screenshots. They help to identify what are the changes in the UI.

My notes:
- The `wp wcpay add_document_for_account` command from the testing instructions did not work for me. I had to add `--date` parameter.

```
wp wcpay add_document_for_account {account_id} vat_invoice --mode=test --date="2024-07-24" --period_from="2023-01-01 00:00:00" --period_to="2024-07-31 23:59:59"
```

- There is a [linter issue](https://github.com/Automattic/woocommerce-payments/actions/runs/10085330545/job/27885923855?pr=8974) that needs to be fixed.

- I tested using US, UK, and JP store.

- Do you plan to resolve some of the TODOs in this PR?

- Acknowledging that GST is mentioned in some of the comments but it's not returned in `getVatTaxIDName()` because GST, according to ChatGPT, for India, Australia, Canada, New Zealand.

- I tested that the hint returned by `getVatTaxIDValidationHint()` for JP works, ie entering `1234567890123` passes the validation.

I didn't think too much about e2e test but if it's not straight forward, better to do in a separate PR. Not sure if setting certain country is easy in e2e test.

brucealdridge reviewed (approved):
I've tested the happy path for JP 🇯🇵 and everything looks good.

The code changes look good to me. When I saw the warning about substrings inside translated strings I was concerned, but the approach here is sensible and good to merge.

haszari commented:
Thanks @brucealdridge  @shendy-a8c ! I'm working on the e2e tests now, I _think_ I can get that sorted, and will also tidy up linter and todos/comments before merge. [Moved back to `In progress`]

Commit:
fix references to (changed) tax modal title in tests

Commit:
remove todo comments, clarify, link to follow up issue for legal/regulatory warning

haszari committed:
Merge branch 'develop' into fix/8926-localise-vat-details-modal

Commit:
fix typo – VAT Number is capitalised in UI

Commit:
parameterise e2e test expectation for tax id name

Commit:
update snapshot for tax details modal to match new code

Commit:
remove obsolete e2e snapshot

Commit:
fix test for no prefix for JP corporate number 😅

haszari commented:
- Tests are sorted now ✅
- Comments & todos tidied 📝

> Acknowledging that GST is mentioned in some of the comments but it's not returned in getVatTaxIDName() because GST, according to ChatGPT, for India, Australia, Canada, New Zealand.

That's by design, a hint for anyone working on this in future to not assume "VAT". We don't yet support tax docs in any "GST" regions (as of right now). 

Merging 🚢

haszari commented:
> Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

I have the ball, will follow up & add to What's New.
